File: src/main/java/com/rdvdev2/timetravelmod/impl/common/command/TimeCorruptionCommand.java
package com.rdvdev2.timetravelmod.impl.common.command;

import com.mojang.brigadier.Command;
import com.mojang.brigadier.CommandDispatcher;
import com.mojang.brigadier.arguments.IntegerArgumentType;
import com.mojang.brigadier.builder.LiteralArgumentBuilder;
import com.mojang.brigadier.context.CommandContext;
import com.mojang.brigadier.exceptions.CommandSyntaxException;
import com.mojang.brigadier.exceptions.DynamicCommandExceptionType;
import com.mojang.brigadier.tree.LiteralCommandNode;
import com.rdvdev2.timetravelmod.api.dimension.ITimeline;
import com.rdvdev2.timetravelmod.impl.common.command.argument.TimelineArgumentType;
import net.minecraft.server.command.CommandManager;
import net.minecraft.server.command.ServerCommandSource;
import net.minecraft.text.LiteralText;
import net.minecraft.text.TranslatableText;

public class TimeCorruptionCommand {

    private static final DynamicCommandExceptionType NO_TIMELINE_FOUND_EXCEPTION;

    public static void register(CommandDispatcher<ServerCommandSource> dispatcher) {
        LiteralCommandNode<ServerCommandSource> node = dispatcher.register(command());
        dispatcher.register(CommandManager.literal("tc").redirect(node));
    }

    private static LiteralArgumentBuilder<ServerCommandSource> command() {
        return CommandManager.literal("timecorruption")
                .requires(c -> c.hasPermissionLevel(2))
                .then(CommandManager.literal("get")
                        .then(CommandManager.argument("timeline", TimelineArgumentType.timeline())
                                .executes(TimeCorruptionCommand::executeGet))
                        .executes(TimeCorruptionCommand::executeGet))
                .then(CommandManager.literal("set")
                        .then(CommandManager.argument("timeline", TimelineArgumentType.timeline())
                                .then(CommandManager.argument("value", IntegerArgumentType.integer(0))
                                    .executes(TimeCorruptionCommand::executeSet))))
                .then(CommandManager.literal("add")
                        .then(CommandManager.argument("timeline", TimelineArgumentType.timeline())
                                .then(CommandManager.argument("value", IntegerArgumentType.integer())
                                        .executes(TimeCorruptionCommand::executeAdd))));

    }

    private static ITimeline getTimeline(CommandContext<ServerCommandSource> commandContext) throws CommandSyntaxException {
        try {
            return TimelineArgumentType.getTimelineArgument(commandContext, "timeline");
        } catch (IllegalArgumentException e) {
            ITimeline timeline = ITimeline.getTimelineForWorld(commandContext.getSource().getWorld().getRegistryKey());
            if (timeline != null) return timeline;
            else throw NO_TIMELINE_FOUND_EXCEPTION.create(commandContext.getSource().getWorld().getRegistryKey().getValue());
        }
    }

    private static int executeGet(CommandContext<ServerCommandSource> commandContext) throws CommandSyntaxException {
        ITimeline timeline = getTimeline(commandContext);
        int value = timeline.getCorruption(commandContext.getSource().getMinecraftServer()).getCorruptionLevel();
        commandContext.getSource().sendFeedback(new TranslatableText("command.time_travel_mod.time_corruption.result", value), false);
        return Command.SINGLE_SUCCESS;
    }

    private static int executeSet(CommandContext<ServerCommandSource> commandContext) throws CommandSyntaxException {
        ITimeline timeline = getTimeline(commandContext);
        int value = IntegerArgumentType.getInteger(commandContext, "value");
        value = timeline.getCorruption(commandContext.getSource().getMinecraftServer()).setCorruptionLevel(value);
        commandContext.getSource().sendFeedback(new TranslatableText("command.time_travel_mod.time_corruption.result", value), false);
        return Command.SINGLE_SUCCESS;
    }

    private static int executeAdd(CommandContext<ServerCommandSource> commandContext) throws CommandSyntaxException {
        ITimeline timeline = getTimeline(commandContext);
        int value = IntegerArgumentType.getInteger(commandContext, "value");
        value = timeline.getCorruption(commandContext.getSource().getMinecraftServer()).increaseCorruptionLevel(value);
        commandContext.getSource().sendFeedback(new TranslatableText("command.time_travel_mod.time_corruption.result", value), false);
        return Command.SINGLE_SUCCESS;
    }

    static {
        NO_TIMELINE_FOUND_EXCEPTION = new DynamicCommandExceptionType( o ->
                new TranslatableText("command.time_travel_mod.time_corruption.error", o));
    }
}

File: src/main/kotlin/com/rdvdev2/timetravelmod/impl/common/command/TimeCorruptionCommand.kt
package com.rdvdev2.timetravelmod.impl.common.command

import com.mojang.brigadier.Command
import com.mojang.brigadier.CommandDispatcher
import com.mojang.brigadier.arguments.IntegerArgumentType
import com.mojang.brigadier.builder.LiteralArgumentBuilder
import com.mojang.brigadier.context.CommandContext
import com.mojang.brigadier.exceptions.CommandSyntaxException
import com.mojang.brigadier.exceptions.DynamicCommandExceptionType
import com.rdvdev2.timetravelmod.api.dimension.ITimeline
import com.rdvdev2.timetravelmod.api.dimension.ITimeline.Companion.getTimelineForWorld
import com.rdvdev2.timetravelmod.impl.common.command.argument.TimelineArgumentType.Companion.getTimelineArgument
import com.rdvdev2.timetravelmod.impl.common.command.argument.TimelineArgumentType.Companion.timeline
import net.minecraft.server.command.CommandManager
import net.minecraft.server.command.ServerCommandSource
import net.minecraft.text.TranslatableText

object TimeCorruptionCommand {

    private val NO_TIMELINE_FOUND_EXCEPTION = DynamicCommandExceptionType { TranslatableText("command.time_travel_mod.time_corruption.error", it) }

    fun register(dispatcher: CommandDispatcher<ServerCommandSource>) {
        val node = dispatcher.register(command())
        dispatcher.register(CommandManager.literal("tc").redirect(node))
    }

    private fun command(): LiteralArgumentBuilder<ServerCommandSource> {
        return CommandManager.literal("timecorruption")
            .requires { it.hasPermissionLevel(2) }
            .then(CommandManager.literal("get")
                .then(CommandManager.argument("timeline", timeline())
                    .executes(this::executeGet)
                )
                .executes(this::executeGet)
            )
            .then(CommandManager.literal("set")
                .then(CommandManager.argument("timeline", timeline())
                    .then(CommandManager.argument("value", IntegerArgumentType.integer(0))
                        .executes(this::executeSet)
                    )
                )
            )
            .then(CommandManager.literal("add")
                .then(CommandManager.argument("timeline", timeline())
                    .then(CommandManager.argument("value", IntegerArgumentType.integer())
                        .executes (this::executeAdd)
                    )
                )
            )
    }

    @Throws(CommandSyntaxException::class)
    private fun getTimeline(commandContext: CommandContext<ServerCommandSource>): ITimeline {
        return try {
            getTimelineArgument(commandContext, "timeline")
        } catch (e: IllegalArgumentException) {
            val timeline = getTimelineForWorld(commandContext.source.world.registryKey)
            timeline ?: throw NO_TIMELINE_FOUND_EXCEPTION.create(commandContext.source.world.registryKey.value)
        }
    }

    @Throws(CommandSyntaxException::class)
    private fun executeGet(commandContext: CommandContext<ServerCommandSource>): Int {
        val timeline = getTimeline(commandContext)
        val value = timeline.getCorruption(commandContext.source.minecraftServer).corruptionLevel
        commandContext.source.sendFeedback(TranslatableText("command.time_travel_mod.time_corruption.result", value), false)
        return Command.SINGLE_SUCCESS
    }

    @Throws(CommandSyntaxException::class)
    private fun executeSet(commandContext: CommandContext<ServerCommandSource>): Int {
        val timeline = getTimeline(commandContext)
        var value = IntegerArgumentType.getInteger(commandContext, "value")
        value = timeline.getCorruption(commandContext.source.minecraftServer).setCorruptionLevel(value)
        commandContext.source.sendFeedback(TranslatableText("command.time_travel_mod.time_corruption.result", value), false)
        return Command.SINGLE_SUCCESS
    }

    @Throws(CommandSyntaxException::class)
    private fun executeAdd(commandContext: CommandContext<ServerCommandSource>): Int {
        val timeline = getTimeline(commandContext)
        var value = IntegerArgumentType.getInteger(commandContext, "value")
        value = timeline.getCorruption(commandContext.source.minecraftServer).increaseCorruptionLevel(value)
        commandContext.source.sendFeedback(TranslatableText("command.time_travel_mod.time_corruption.result", value), false)
        return Command.SINGLE_SUCCESS
    }

}